disktable.go

```diff
@@ -0,0 +1,650 @@
+	primaryPath := filepath.Join(pathDir, "primary")
+	badgeOpts := readOpts.WithDir(primaryPath).WithValueDir(primaryPath).WithLogger(table.logger)
+	db, err := badger.Open(badgeOpts)
+	if err != nil {
+		return nil, err
+	}
+	table.primary = &Index{
+		Name:     "primary",
+		db:       db,
+		buffPool: make(chan *z.Buffer, 100),
+		counter:  counter,
+	}
+
+	return table, nil
+}
+
+// Get gets the i'th entry stored in the table.
+func (t *Table) Get(ctx context.Context, i uint64) ([]byte, error) {
+	// The primary index is stored with the starting index as 1. Externally
+	// we use the standard starting index of 0. So we i++.
+	i++
+
+	var v []byte
+	err := t.primary.db.View(
+		func(txn *badger.Txn) error {
+			item, err := txn.Get(NumToByte(i))
+			if err != nil {
+				return err
+			}
+
+			v, err = item.ValueCopy(nil)
+			if err != nil {
+				return fmt.Errorf("problem getting value at index %d: %w", i, err)
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// Len() returns the number of entries in the table.
+func (t *Table) Len() uint64 {
+	return t.primary.counter.Load()
+}
+
+// Lookup provides the Index name and the Value that needs to match for the entry
+// to be returned.
+type Lookup struct {
+	// IndexName is the name of the index to do the lookup in.
+	IndexName string
+	// Key is the key in the index to lookup.
+	Key []byte
+}
+
+// Result is the result of a table lookup.
+type Result struct {
+	Value []byte
+	Err   error
+}
+
+type btreeItem struct {
+	primaryKey uint64
+	count      uint16
+}
+
+func (a btreeItem) Less(than btree.Item) bool {
+	b := than.(btreeItem)
+	return a.primaryKey < b.primaryKey
+}
+
+// Fetch retrieves specifc rows that match all index lookups. You must supply at least
+// 1 lookup. You cannot currently specify multiple searches in the same index.
+// If you wish to fetch all rows, use FetchAll().
+func (t *Table) Fetch(ctx context.Context, lookup ...Lookup) (chan Result, error) {
+	indexesSearched := map[string]bool{}
+
+	for _, l := range lookup {
+		if t.indexByName[l.IndexName] == nil {
+			return nil, fmt.Errorf("index %q is not found", l.IndexName)
+		}
+		if indexesSearched[l.IndexName] {
+			return nil, fmt.Errorf("index %q cannot be used in a search more than once", l.IndexName)
+		}
+		indexesSearched[l.IndexName] = true
+	}
+
+	if len(lookup) == 0 {
+		return nil, fmt.Errorf("Fetch() requires lookups, otherwise use FetchAll()")
+	}
+
+	ch := make(chan Result, 1)
+
+	go func() {
+		defer close(ch)
+
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		wg := sync.WaitGroup{}
+		mu := sync.Mutex{}
+		matches := btree.New(2)
+
+		// fetch all rows for each index that match our lookup constraint.
+		// Record those rows into a 2-3-4 tree. Any entry in the tree that has
+		// len(lookup) as its value is a line we want to return.
+		errCh := make(chan error, 1)
+		for _, l := range lookup {
+			l := l
+
+			wg.Add(1)
+			go func() {
```

Comment:
just a style nit - it usually makes more sense (to me) to pass `l` into the anonymous function:

```suggestion
			wg.Add(1)
			go func(l Lookup) {
```

(and then obviously passing `l` to the function invocation below)

Reply:
With the change to "golang.org/x/sync/errgroup", this is no longer relevant. See below.